Check if the owner of a share exists

rullzer opened the pull request:
If the owner does not exist we should behave properly.

We can't delete the share at this point as a back end might just be temp
offline.


To test:

1. Setup NC with LDAP
2. Share a folder `test` from user1 to user2
3. Delete user1 from LDAP
  * DO NOT NOTIFY NEXTCLOUD
4. Login as user2
5. navigate to `test`

Before:
* If you enter test you will not see any content and warnings will be thrown
* Weird sync behavior

Now:
* Warning that the storage is temporaly unavailable
* Sync client skips the folder because it is not available

blizzz reviewed (dismissed):
tested

schiessle commented:
I remember that I discussed this issue with @blizzz a few days/weeks ago. If I remember correctly our conclusion was that we should still show the folder but in a "temporarily not available" state, same as we do for external storage. If we just don't show it at all this will trigger the sync client to delete it and if the user back-end comes back the sync client will have to download everything again. Which can be really painful for large shared folders.

MorrisJobke commented:
> I remember that I discussed this issue with @blizzz a few days/weeks ago because I saw a similar issue on another instance. If I remember correctly our conclusion was that we should still show the folder but in a "temporarily not available" state, same as we do for external storage. If we just don't show it at all this will trigger the sync client to delete it and if the user back-end comes back the sync client will have to download everything again. Which can be really painful for large shared folders.

AFAIK this is done via a `StorageNotAvailableException` right, @icewind1991 ?

icewind1991 commented:
> AFAIK this is done via a StorageNotAvailableException right, @icewind1991 ?

Yes.

rullzer commented:
But where to throw it :P??

rullzer committed:
A failed storage is a not available storage

We have to double check. Since getting the info of the root returns a
generic entry. But actually the stroage is not available. Else we get
very weird sync and web behavior.

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
@icewind1991 please see my two commits. I think they make it somewhat more sane. As the FailedStorage was not handle properly by the DAV app. Now it behaves better IMO.

@schiessle The making red of the storage is actually handled by the files_external javascript.

rullzer commented:
Ok, so my initial commit did not make much sense. As LDAP will also find out that the user is deleted. And still return it.

Now the share will still show. But if you try to enter it you get a 'storage not available' notification. This is probably saner.

brunt82 commented:
One short question: how can the client request the server to get the current folder structure, if it is unable to log in, because the user backend is not available?

Do you have any use case where one user is able to login but another user not because it is temporarly disabled?

blizzz commented:
@brunt82 no, LDAP is available. This case is about that a user disappeared from LDAP who was sharing a file or folder to another one, who logged in. And sees the shares, but has no way to access them and experiences confusing behaviour.

rullzer committed:
Fix tests

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
Tests fixed

icewind1991 commented on apps/files_sharing/lib/SharedStorage.php:
```diff
@@ -357,6 +359,8 @@ public function getItemType() {
 	 * @return Cache
 	 */
 	public function getCache($path = '', $storage = null) {
+		$this->init();
```
this causes a major performance hit.

Shared storage is left uninitialized for cache only work for a reason

rullzer replied:
This is already the case. 

https://github.com/nextcloud/server/blob/master/apps/files_sharing/lib/SharedStorage.php#L369
https://github.com/nextcloud/server/blob/master/apps/files_sharing/lib/Cache.php#L72
https://github.com/nextcloud/server/blob/master/apps/files_sharing/lib/SharedStorage.php#L458
https://github.com/nextcloud/server/blob/master/apps/files_sharing/lib/SharedStorage.php#L459

This just moves it earlier so the the cache of the storage is also set to FailedCache.

icewind1991 replied:
Will need to fix that then...

icewind1991 replied:
https://github.com/nextcloud/server/pull/7385

rullzer replied:
Fair enough then let me drop this commit from here. As the first one is the most important one anyway I think

LukasReschke reviewed (approved):
🐘
